Assignments.java

```diff
@@ -0,0 +1,143 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Assignments {
+
+	public static void main(String[] args) {
+		Assignment1 assignment = new Assignment1();
+
+		System.out.println(assignment.q1Permutation("Apple", "Pabble"));
+
+		Node<Integer> n1 = new Node<>(7);
+		Node<Integer> n2 = new Node<>(6, n1);
+		Node<Integer> n3 = new Node<>(5, n2);
+		Node<Integer> n4 = new Node<>(4, n3);
+		Node<Integer> n5 = new Node<>(3, n4);
+		Node<Integer> n6 = new Node<>(2, n5);
+		Node<Integer> n7 = new Node<>(1, n6);
+
+		System.out.println(assignment.q2KthToLast(n7, 5));
+	}
+
+	static class Assignment1 {
+
+		/**
+		 * Checks if two strings are permutations of each other
+		 * 
+		 * @param s1
+		 *            String1
+		 * @param s2
+		 *            String2
+		 * @return true if s1 and s2 are permutations, false otherwise
+		 */
+		public boolean q1Permutation(String s1, String s2) {
+			if (s1 == null || s2 == null) {
+				return false;
+			}
+
+			// to count A and a as the same letter
+			s1 = s1.toLowerCase();
+			s2 = s2.toLowerCase();
+
+			// if s1 and s2's lengths are not equal, they can't be permutations
+			if (s1.length() != s2.length()) {
+				return false;
+			}
+
+			Map<Character, Integer> map = new HashMap<>();
+
+			// maps each character of s1 to its number of occurrences in s1
+			for (Character c : s1.toCharArray()) {
+				int occurrenceBefore = map.getOrDefault(c, 0);
+				map.put(c, occurrenceBefore + 1);
+			}
+
+			for (Character c : s2.toCharArray()) {
+				// if c is not present in s1, s1 and s2 are not permutations
+				if (!map.containsKey(c)) {
+					return false;
+				}
+				// decreases value of c by one whenever it is found
+				map.put(c, map.get(c) - 1);
+
+				// if value of c is less than zero, s1 and s2 are not
+				// permutations
+				if (map.get(c) < 0) {
+					return false;
+				}
+			}
+
+			return true;
+		}
+
+		/**
+		 * Finds kth to last item in a singly linkedlist.
+		 * 
+		 * @param list
+		 *            Head node of the singly linkedlist
+		 * @param k
+		 *            Index from last of the node to be returned
+		 * @return kth to last element
+		 */
+		public Integer q2KthToLast(Node<Integer> list, int k) {
+			if (list == null || k < 0) {

```

Comment:
Another possible way of handling k<0 (and p1==null in line 95) would be by throwing IllegalArgumentException.

Both exception and returning null are valid, but because there is more than one option available, it is usually a good idea to explicitly describe behavior in method description, for example "@return kth to last element or null if k is out of valid range".